runtime/alloc/collect.c

```diff
@@ -70,29 +78,40 @@ static void migrate_once(block** blockPtr) {
   block* currBlock = *blockPtr;
   memory_block_header *hdr = mem_block_header(currBlock);
   // bit has been flipped by now, so we need != and not ==
-  if (true_is_fromspace != hdr->semispace) {
+  if (fromspace_id() == hdr->semispace) {
     migrate(blockPtr);
   }
 }
 
 static void migrate_string_buffer(stringbuffer** bufferPtr) {
```

Comment:
Buffers don't have a separate header from the header of their contents. As I said, the current implementation treats them as one object, not two.

Reply:
Even if we try to move them together, isn't the buffer a separate block from the contents? At the very least the buffer object rather than the contents should hold the forwarding pointer after migration.

Reply:
as I said, they're treated atomically by the current implementation. the stringbuffer doesn't have a separate header, it just usese the header of the underlying string, so it can't have a forwarding pointer. This could maybe be changed, I suppose, but I don't see the benefit.